Extend logging capabilities

andreas-bok-sociomantic opened the pull request:
Add filename, line number and funcname to log output

andreas-bok-sociomantic committed:
Define auxillary macros for logging

Defines macros for logging with file, line number
and function name

andreas-bok-sociomantic committed:
Add filename, line number and function name to log output

Rewrite log function so filename, line number function name
are logged as well. The format is

[LEVEL] [filename:linenumber] : functionname - logmessage
e.g.
DEBUG [libdrizzle/drizzle.cc:351] : drizzle_wait - poll return=1

Use the new logging Macros to add the extra parameters to
the calls to the log function

andreas-bok-sociomantic committed:
Add logging where missing

andreas-bok-sociomantic committed:
Modify the client callback log function signature

Add filename, line number and function name to the
signature of callback function.

andreas-bok-sociomantic committed:
Add client log callback function to connect unittest

andreas-bok-sociomantic committed:
Add file, line, function params to drizzle_set_error

andreas-bok-sociomantic committed:
Add file, line, function  to connect_failed_try_next

ben-palmer-sociomantic commented on libdrizzle/drizzle_local.h:
```diff
@@ -52,6 +52,12 @@ extern "C" {
+ * Macros for logging with file, line number and function name
+ */
+#define __FILE_LINE_FUNC__ __FILE__,__LINE__, __func__
+#define __LOG_FUNCTION__ drizzle_log_debug(con, __FILE_LINE_FUNC__, "");
```
I might be tempted to change this to `__LOG_LOCATION__` as that is really what it is doing but its just an idea.

andreas-bok-sociomantic replied:
Good point. I will change it

ben-palmer-sociomantic commented on libdrizzle/binlog.cc:
```diff
@@ -473,15 +473,17 @@ void drizzle_binlog_get_filename(drizzle_st *con, char **filename, int file_inde
   if (driz_ret != DRIZZLE_RETURN_OK)
   {
-    drizzle_log_error(con, "%s Couldn't retrieve  %d", __func__, file_index );
+    drizzle_log_error(con, __FILE_LINE_FUNC__,
```
There is a disconnect in the commit message between printing the file, line, and function and what you give as an example: `DEBUG [libdrizzle/drizzle.cc:351] : drizzle_wait poll return=1`. It seems to be missing the function name.

ben-palmer-sociomantic commented on libdrizzle/state.cc:
```diff
@@ -100,8 +100,9 @@ drizzle_return_t drizzle_state_packet_read(drizzle_st *con)
-    ", packet_number= %" PRIu8, con->buffer_size, con->packet_size, con->packet_number);
+  drizzle_log_debug(con, __FILE_LINE_FUNC__,
+    "buffer_size= %" PRIu64 ", packet_size= %" PRIu32 ", packet_number= %" PRIu8,
+    sizeof(size_t), con->buffer_size, con->packet_size, con->packet_number);
```
Where did `sizeof(size_t)` come from here?

andreas-bok-sociomantic replied:
Not sure, but definitely shouldn't be there

ben-palmer-sociomantic commented on tests/unit/connect.c:
```diff
@@ -37,11 +37,25 @@
+#include "tests/unit/common.c"
 #include <libdrizzle-5.1/libdrizzle.h>
 #include <stdint.h>
+
```
Extra new line here.

ben-palmer-sociomantic commented on tests/unit/connect.c:
```diff
@@ -37,11 +37,25 @@
 #include <stdlib.h>
 
+extern void log_fn_callback(const char *file, uint line, const char *func,
+  const char *msg, drizzle_verbose_t verbose, void *context );
```
Extra white space between `context` and `);`

ben-palmer-sociomantic commented on tests/unit/connect.c:
```diff
@@ -50,12 +64,12 @@ int main(int argc, char *argv[])
-                                  getenv("MYSQL_USER"),
-                                  getenv("MYSQL_PASSWORD"),
-                                  getenv("MYSQL_SCHEMA"), opts);
+  con= drizzle_create(getenv("MYSQL_SERVER"),
```
Should these whitespace changes be part of this commit?

andreas-bok-sociomantic replied:
Indirectly since with import of `common.c` the `con` variable is now defined there. So the removal of the type changes the indentation.

ben-palmer-sociomantic commented on libdrizzle/binlog.cc:
```diff
@@ -337,8 +337,8 @@ drizzle_return_t drizzle_state_binlog_read(drizzle_st *con)
     {
-        drizzle_set_error(con, __func__,
-                          "packet size error:%" PRIu32 ":%" PRIu32, con->packet_size, binlog_event->length);
+        drizzle_set_error(con, __FILE_LINE_FUNC__, "packet size error:%" PRIu32 \
```
I might be wrong here but is the use of `\` to indicate multiple lines inconsistent with how this is done in other modules?

ben-palmer-sociomantic commented on libdrizzle/conn.cc:
```diff
@@ -1354,7 +1354,7 @@ drizzle_return_t drizzle_state_read(drizzle_st *con)
       if (con->flags.is_shutdown == false)
       {
-        drizzle_set_error(con, __func__,
+        drizzle_set_error(con, __FILE_LINE_FUNC__,
```
The file and line included in the message are now redundant.

ben-palmer-sociomantic commented on libdrizzle/conn.cc:
```diff
@@ -1415,14 +1415,14 @@ drizzle_return_t drizzle_state_read(drizzle_st *con)
       case ECONNRESET:
         {
-          drizzle_set_error(con, __func__,
+          drizzle_set_error(con, __FILE_LINE_FUNC__,
```
Again here, might pay to check them all.

ben-palmer-sociomantic reviewed (commented):
I have done an initial review of this PR. I like the way this seems to make the logging and error function consistent.

ben-palmer-sociomantic commented on libdrizzle/conn.cc:
```diff
@@ -1514,12 +1512,12 @@ drizzle_return_t drizzle_state_write(drizzle_st *con)
       {
-        drizzle_set_error(con, __func__, "%s:%d lost connection to server (%s)",
-                          __FILE__, __LINE__, strerror(errno));
+        drizzle_set_error(con, __FILE_LINE_FUNC__, "%s:%d lost connection to server (%s)",
```
Is this format incorrect now we have removed `__FILE__` and `__LINE__`

andreas-bok-sociomantic replied:
oops, no wrong. will fix

ben-palmer-sociomantic commented on libdrizzle/drizzle_local.h:
```diff
@@ -65,7 +65,8 @@ extern "C" {
  * @param[in] format Format and variable argument list of message.
  */
-void drizzle_set_error(drizzle_st *con, const char *function, const char *format, ...);
+void drizzle_set_error(drizzle_st *con, const char *file, uint line,
```
Should the comments above also be updated.

ben-palmer-sociomantic commented:
I had two comments, otherwise LGTM.

andreas-bok-sociomantic commented:
Addressed comments from ben

```diff
diff --git a/libdrizzle/column.cc b/libdrizzle/column.cc
index 461d15a..c564e6a 100644
--- a/libdrizzle/column.cc
+++ b/libdrizzle/column.cc
@@ -396,7 +396,7 @@ drizzle_return_t drizzle_column_buffer(drizzle_result_st *result)
     result->column_buffer= new (std::nothrow) drizzle_column_st[result->column_count];
     if (result->column_buffer == NULL)
     {
-      drizzle_set_error(result->con,  __FILE_LINE_FUNC__, "Failed to allocate.");
+      drizzle_set_error(result->con, __FILE_LINE_FUNC__, "Failed to allocate.");
 
       return DRIZZLE_RETURN_MEMORY;
     }
diff --git a/libdrizzle/conn.cc b/libdrizzle/conn.cc
index 34d339c..5fb7fc4 100644
--- a/libdrizzle/conn.cc
+++ b/libdrizzle/conn.cc
@@ -821,8 +821,7 @@ drizzle_result_st *drizzle_command_write(drizzle_st *con,
   {
     if (con->state.raw_packet)
     {
-      drizzle_set_error(con, __FILE_LINE_FUNC__,
-                        "connection not ready");
+      drizzle_set_error(con, __FILE_LINE_FUNC__, "connection not ready");
       *ret_ptr= DRIZZLE_RETURN_NOT_READY;
       return result;
     }
@@ -1355,7 +1354,7 @@ drizzle_return_t drizzle_state_read(drizzle_st *con)
     {
       if (con->flags.is_shutdown == false)
       {
-        drizzle_set_error(con, __FILE_LINE_FUNC__, "%s:%d lost connection to server (EOF)");
+        drizzle_set_error(con, __FILE_LINE_FUNC__, "lost connection to server (EOF)");
       }
 
       return DRIZZLE_RETURN_LOST_CONNECTION;
@@ -1416,7 +1415,7 @@ drizzle_return_t drizzle_state_read(drizzle_st *con)
       case ECONNRESET:
         {
           drizzle_set_error(con, __FILE_LINE_FUNC__,
-                            "%s:%d lost connection to server (%s)", strerror(errno));
+                            "lost connection to server (%s)", strerror(errno));
           return DRIZZLE_RETURN_LOST_CONNECTION;
         }
       }
@@ -1513,7 +1512,7 @@ drizzle_return_t drizzle_state_write(drizzle_st *con)
       }
       else if (errno == EPIPE || errno == ECONNRESET)
       {
-        drizzle_set_error(con, __FILE_LINE_FUNC__, "%s:%d lost connection to server (%s)",
+        drizzle_set_error(con, __FILE_LINE_FUNC__, "lost connection to server (%s)",
                           strerror(errno));
         return DRIZZLE_RETURN_LOST_CONNECTION;
       }
diff --git a/libdrizzle/drizzle_local.h b/libdrizzle/drizzle_local.h
index b8af98f..a1dd92b 100644
--- a/libdrizzle/drizzle_local.h
+++ b/libdrizzle/drizzle_local.h
@@ -61,7 +61,9 @@ extern "C" {
 /**
  * Set the error string.
  *
- * @param[in] drizzle Drizzle con structure
+ * @param[in] con Drizzle con structure
+ * @param[in] file Name of file the error happened in.
+ * @param[in] line Linenumber where the error happened.
  * @param[in] function Name of function the error happened in.
  * @param[in] format Format and variable argument list of message.
  */
diff --git a/libdrizzle/field.cc b/libdrizzle/field.cc
index 92eefc9..4702014 100644
--- a/libdrizzle/field.cc
+++ b/libdrizzle/field.cc
@@ -150,7 +150,7 @@ drizzle_field_t drizzle_field_buffer(drizzle_result_st *result, size_t *total,
 #if SIZE_MAX < UINT64_MAX
   if (wire_size >= SIZE_MAX)
   {
-    drizzle_set_error(result->con,  __FILE_LINE_FUNC__, "Field is larger than memory.");
+    drizzle_set_error(result->con, __FILE_LINE_FUNC__, "Field is larger than memory.");
     *ret_ptr= DRIZZLE_RETURN_MEMORY;
     return NULL;
   }
@@ -200,7 +200,7 @@ drizzle_field_t drizzle_field_buffer(drizzle_result_st *result, size_t *total,
     result->field_buffer_sizes[current_field]= (*total) + 1;
     if (result->field_buffer[current_field] == NULL)
     {
-      drizzle_set_error(result->con,  __FILE_LINE_FUNC__, "Failed to allocate.");
+      drizzle_set_error(result->con, __FILE_LINE_FUNC__, "Failed to allocate.");
       *ret_ptr= DRIZZLE_RETURN_MEMORY;
       return NULL;
     }
diff --git a/libdrizzle/pack.cc b/libdrizzle/pack.cc
index f3af4f8..0bc3094 100644
--- a/libdrizzle/pack.cc
+++ b/libdrizzle/pack.cc
@@ -320,8 +320,7 @@ drizzle_return_t drizzle_unpack_string(drizzle_st *con, char *buffer,
   {
     if (ret == DRIZZLE_RETURN_NULL_SIZE)
     {
-      drizzle_set_error(con, __FILE_LINE_FUNC__,
-                        "unexpected NULL length");
+      drizzle_set_error(con, __FILE_LINE_FUNC__, "unexpected NULL length");
     }
 
     return ret;
@@ -329,8 +328,7 @@ drizzle_return_t drizzle_unpack_string(drizzle_st *con, char *buffer,
 
   if (length > con->packet_size)
   {
-    drizzle_set_error(con, __FILE_LINE_FUNC__,
-                           "string extends past end of packet");
+    drizzle_set_error(con, __FILE_LINE_FUNC__, "string extends past end of packet");
     return DRIZZLE_RETURN_UNEXPECTED_DATA;
   }
   if (length > con->buffer_size)
@@ -451,8 +449,7 @@ static drizzle_return_t _pack_scramble_hash(drizzle_st *con,
 
   if (con->scramble == NULL)
   {
-    drizzle_set_error(con, __FILE_LINE_FUNC__,
-                      "no scramble buffer");
+    drizzle_set_error(con, __FILE_LINE_FUNC__, "no scramble buffer");
     return DRIZZLE_RETURN_NO_SCRAMBLE;
   }
 
@@ -498,7 +495,7 @@ bool drizzle_check_unpack_error(drizzle_st *con)
   snprintf(con->last_error, DRIZZLE_MAX_ERROR_SIZE, "%.*s",
            (int)con->packet_size-1, con->buffer_ptr);
 
-  drizzle_set_error(con, __func__, " %s", con->last_error);
+  drizzle_set_error(con, __FILE_LINE_FUNC__, " %s", con->last_error);
 
   con->buffer_ptr+= con->packet_size;
   con->buffer_size-= con->packet_size;
diff --git a/libdrizzle/result.cc b/libdrizzle/result.cc
index ce6f858..c85ab69 100644
--- a/libdrizzle/result.cc
+++ b/libdrizzle/result.cc
@@ -339,7 +339,7 @@ drizzle_return_t drizzle_result_buffer(drizzle_result_st *result)
       if (row_list == NULL)
       {
         drizzle_row_free(result, row);
-        drizzle_set_error(result->con,  __FILE_LINE_FUNC__, "Failed to realloc row_list.");
+        drizzle_set_error(result->con, __FILE_LINE_FUNC__, "Failed to realloc row_list.");
         return DRIZZLE_RETURN_MEMORY;
       }
       result->row_list= row_list;
@@ -350,7 +350,7 @@ drizzle_return_t drizzle_result_buffer(drizzle_result_st *result)
         if (null_bitmap_list == NULL)
         {
           drizzle_row_free(result, row);
-          drizzle_set_error(result->con,  __FILE_LINE_FUNC__, "Failed to realloc null_bitmap_list.");
+          drizzle_set_error(result->con, __FILE_LINE_FUNC__, "Failed to realloc null_bitmap_list.");
           return DRIZZLE_RETURN_MEMORY;
         }
         result->null_bitmap_list= null_bitmap_list;
diff --git a/libdrizzle/statement.cc b/libdrizzle/statement.cc
index b62cf52..8101590 100644
--- a/libdrizzle/statement.cc
+++ b/libdrizzle/statement.cc
@@ -119,7 +119,7 @@ drizzle_return_t drizzle_stmt_execute(drizzle_stmt_st *stmt)
   {
     if (!stmt->query_params[current_param].is_bound)
     {
-      drizzle_set_error(stmt->con,  __FILE_LINE_FUNC__, "parameter %d has not been bound", current_param);
+      drizzle_set_error(stmt->con, __FILE_LINE_FUNC__, "parameter %d has not been bound", current_param);
       return DRIZZLE_RETURN_STMT_ERROR;
     }
     /* Don't count NULLs */
@@ -143,7 +143,7 @@ drizzle_return_t drizzle_stmt_execute(drizzle_stmt_st *stmt)
   buffer = new (std::nothrow) unsigned char[buffer_size];
   if (buffer == NULL)
   {
-    drizzle_set_error(stmt->con,  __FILE_LINE_FUNC__, "new");
+    drizzle_set_error(stmt->con, __FILE_LINE_FUNC__, "new");
     return DRIZZLE_RETURN_MEMORY;
   }
   buffer_pos= buffer;
@@ -274,7 +274,7 @@ drizzle_return_t drizzle_stmt_execute(drizzle_stmt_st *stmt)
       case DRIZZLE_COLUMN_TYPE_DATETIME2:
       case DRIZZLE_COLUMN_TYPE_TIME2:
       default:
-        drizzle_set_error(stmt->con,  __FILE_LINE_FUNC__, "unknown type when filling buffer");
+        drizzle_set_error(stmt->con, __FILE_LINE_FUNC__, "unknown type when filling buffer");
         delete[] buffer;
         return DRIZZLE_RETURN_UNEXPECTED_DATA;
         break;
@@ -331,7 +331,7 @@ drizzle_return_t drizzle_stmt_send_long_data(drizzle_stmt_st *stmt, uint16_t par
 
   if (stmt->state < DRIZZLE_STMT_PREPARED)
   {
-    drizzle_set_error(stmt->con,  __FILE_LINE_FUNC__, "stmt object has not been prepared");
+    drizzle_set_error(stmt->con, __FILE_LINE_FUNC__, "stmt object has not been prepared");
     return DRIZZLE_RETURN_STMT_ERROR;
   }
 
@@ -401,7 +401,7 @@ drizzle_return_t drizzle_stmt_fetch(drizzle_stmt_st *stmt)
 
   if (stmt->state < DRIZZLE_STMT_EXECUTED)
   {
-    drizzle_set_error(stmt->con,  __FILE_LINE_FUNC__, "statement has not been executed");
+    drizzle_set_error(stmt->con, __FILE_LINE_FUNC__, "statement has not been executed");
     return DRIZZLE_RETURN_UNEXPECTED_DATA;
   }
   stmt->con->result= stmt->execute_result;
@@ -515,7 +515,7 @@ drizzle_return_t drizzle_stmt_fetch(drizzle_stmt_st *stmt)
         case DRIZZLE_COLUMN_TYPE_DATETIME2:
         case DRIZZLE_COLUMN_TYPE_TIME2:
         default:
-          drizzle_set_error(stmt->con,  __FILE_LINE_FUNC__, "Unknown data type found");
+          drizzle_set_error(stmt->con, __FILE_LINE_FUNC__, "Unknown data type found");
           ret= DRIZZLE_RETURN_UNEXPECTED_DATA;
           break;
       }
@@ -544,7 +544,7 @@ drizzle_return_t drizzle_stmt_buffer(drizzle_stmt_st *stmt)
   }
   if (stmt->state >= DRIZZLE_STMT_FETCHED)
   {
-    drizzle_set_error(stmt->con,  __FILE_LINE_FUNC__, "data set has already been read");
+    drizzle_set_error(stmt->con, __FILE_LINE_FUNC__, "data set has already been read");
     return DRIZZLE_RETURN_UNEXPECTED_DATA;
   }
 
diff --git a/libdrizzle/statement_param.cc b/libdrizzle/statement_param.cc
index d64be0b..3b930d7 100644
--- a/libdrizzle/statement_param.cc
+++ b/libdrizzle/statement_param.cc
@@ -51,7 +51,7 @@ drizzle_return_t drizzle_stmt_set_param(drizzle_stmt_st *stmt, uint16_t param_nu
   }
   if (stmt->state < DRIZZLE_STMT_PREPARED)
   {
-    drizzle_set_error(stmt->con,  __FILE_LINE_FUNC__, "stmt object has not been prepared");
+    drizzle_set_error(stmt->con, __FILE_LINE_FUNC__, "stmt object has not been prepared");
     return DRIZZLE_RETURN_STMT_ERROR;
   }
 ```

andreas-bok-sociomantic commented:
ping

ben-palmer-sociomantic commented:
LGTM

andreas-bok-sociomantic commented:
TX :)
